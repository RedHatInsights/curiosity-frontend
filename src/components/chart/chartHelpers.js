import React from 'react';
import _cloneDeep from 'lodash/cloneDeep';
import { helpers } from '../../common';

/**
 * Generate max X and Y values from datasets.
 *
 * @param {object} params
 * @param {Array} params.dataSets
 * @returns {{individualMaxY: object, maxY: number, maxX: number}}
 */
const generateMaxXY = ({ dataSets = [] } = {}) => {
  const individualDataSetsMaxY = {};
  let combinedDataSetMaxX = 0;
  let combinedDataSetsMaxY = 0;

  dataSets
    .filter(({ isStacked }) => isStacked === true)
    .forEach(({ data }) => {
      if (Array.isArray(data)) {
        combinedDataSetsMaxY += Math.max(...data.map(value => value?.y ?? 0));
      }
    });

  dataSets.forEach(({ id, data }) => {
    let dataSetMaxY = 0;

    if (Array.isArray(data)) {
      combinedDataSetMaxX = data.length > combinedDataSetMaxX ? data.length : combinedDataSetMaxX;

      dataSetMaxY = Math.max(...data.map(value => value?.y ?? 0));
      combinedDataSetsMaxY = dataSetMaxY > combinedDataSetsMaxY ? dataSetMaxY : combinedDataSetsMaxY;
    }

    if (id) {
      individualDataSetsMaxY[id] = dataSetMaxY;
    }
  });

  return {
    maxX: combinedDataSetMaxX,
    maxY: combinedDataSetsMaxY,
    individualMaxY: individualDataSetsMaxY
  };
};

/**
 * Generate Y axis domain ranges from dataSets, ignore X axis.
 *
 * @param {object} params
 * @param {number|object} params.maxY
 * @returns {{ domain: { y: Array } }}
 */
const generateDomains = ({ maxY } = {}) => {
  const updatedChartDomain = {};
  const generatedDomain = {};

  if (Object.values(maxY).length) {
    generatedDomain.y = [0, 1.25];
  } else {
    const floored = Math.pow(10, Math.floor(Math.log10(maxY || 10)));
    generatedDomain.y = [0, Math.ceil((maxY + 1) / floored) * floored];
  }

  if (Object.keys(generatedDomain).length) {
    updatedChartDomain.domain = generatedDomain;
  }

  return {
    ...updatedChartDomain
  };
};

/**
 * Generate chart element props.
 *
 * @param {object} params
 * @param {Array} params.dataSets
 * @param {number} params.maxX
 * @param {number} params.maxY
 * @param {Function} params.xValueFormat
 * @param {Function} params.yValueFormat
 * @returns {{elementsById: object, stackedElements: Array, stackedElementsById: object, elements: Array}}
 */
const generateElementsProps = ({ dataSets = [], maxX, maxY, xValueFormat, yValueFormat }) => {
  const elements = [];
  const stackedElements = [];
  const elementsById = {};
  const stackedElementsById = {};

  dataSets.forEach(dataSet => {
    const {
      animate,
      chartType,
      data,
      fill,
      id,
      isStacked,
      interpolation,
      stroke,
      strokeDasharray,
      strokeWidth
    } = dataSet;

    if (data?.length) {
      const dataColorStroke = {
        data: {}
      };

      if (fill && chartType !== 'line' && chartType !== 'threshold') {
        dataColorStroke.data.fill = fill;
      }

      if (stroke) {
        dataColorStroke.data.stroke = stroke;
      }

      if (strokeDasharray) {
        dataColorStroke.data.strokeDasharray = strokeDasharray;
      }

      if (strokeWidth) {
        dataColorStroke.data.strokeWidth = strokeWidth;
      }

      const defaultProps = {};

      if (animate) {
        defaultProps.animate = animate;
      }

      if (interpolation) {
        defaultProps.interpolation = interpolation;
      }

      const chartElementProps = {
        ...defaultProps,
        key: `chart-${dataSet.id}-${chartType || ''}`,
        name: `chart-${dataSet.id}-${chartType || ''}`,
        data: dataSet.data,
        style: { ...(dataSet.style || {}), ...dataColorStroke },
        themeColor: dataSet.themeColor,
        themeVariant: dataSet.themeVariant,
        x: (xValueFormat && (datum => xValueFormat({ datum, maxX }))) || undefined,
        y:
          (yValueFormat &&
            (datum =>
              yValueFormat({
                datum,
                isMultiAxis: typeof maxY !== 'number',
                maxY: typeof maxY === 'number' ? maxY : maxY?.[dataSet.id]
              }))) ||
          (datum => (typeof maxY === 'number' ? datum.y : datum.y / maxY?.[dataSet.id]))
      };

      const props = { ...chartElementProps };
      const updatedProps = { chartType, props };

      if (isStacked) {
        stackedElementsById[id] = updatedProps;
        stackedElements.push(updatedProps);
      } else {
        elementsById[id] = updatedProps;
        elements.push(updatedProps);
      }
    }
  });

  return {
    elements,
    elementsById,
    stackedElements,
    stackedElementsById
  };
};

/**
 * Preprocess datasets for tooltips.
 *
 * @param {object} params
 * @param {Node|Function} params.content
 * @param {Array} params.dataSets
 * @returns {{}}
 */
const generateTooltipData = ({ content = helpers.noop, dataSets = [] } = {}) => {
  const tooltipDataSetLookUp = {};

  if (content && dataSets?.[0]?.data) {
    dataSets[0].data.forEach((dataSet, index) => {
      const itemsByKey = {};

      dataSets.forEach(data => {
        if (data?.data[index]) {
          itemsByKey[data.id] = {
            color: data.stroke || data.fill || data.color || '',
            chartType: data.chartType,
            data: _cloneDeep(data.data[index])
          };
        }
      });

      const mockDatum = {
        datum: { x: dataSet.x, y: dataSet.y, index, itemsByKey }
      };

      tooltipDataSetLookUp[dataSet.x] = {
        x: dataSet.x,
        y: null,
        itemsByKey,
        tooltip:
          (React.isValidElement(content) && React.cloneElement(content, { ...mockDatum })) || content({ ...mockDatum })
      };
    });
  }

  return tooltipDataSetLookUp;
};

/**
 * Generate X axis props, ticks, tick formatting.
 *
 * @param {object} params
 * @param {object} params.dataSet
 * @param {number} params.maxX
 * @param {number} params.xAxisLabelIncrement
 * @param {object} params.xAxisPropDefaults
 * @param {Function} params.xAxisTickFormat
 * @returns {{tickFormat: (function(*)), tickValues: *}}
 */
const generateXAxisProps = ({
  dataSet = {},
  maxX,
  xAxisLabelIncrement,
  xAxisPropDefaults = {},
  xAxisTickFormat
} = {}) => {
  const { data = [] } = dataSet;
  const axisProps = {
    ...xAxisPropDefaults,
    tickValues: data.reduce(
      (acc, current, index) => (index % xAxisLabelIncrement === 0 ? acc.concat(current.x) : acc),
      []
    ),
    tickFormat: tick => data[tick]?.xAxisLabel || tick
  };

  if (typeof xAxisTickFormat === 'function') {
    axisProps.tickFormat = tick => {
      const tickIndex = axisProps.tickValues.indexOf(tick);
      const previousItem = { ...data[axisProps.tickValues[tickIndex - 1]] };
      const nextItem = { ...data[axisProps.tickValues[tickIndex + 1]] };
      const item = { ...data[tick] };

      return xAxisTickFormat({ tick, previousItem, item, nextItem, maxX });
    };
  }

  return axisProps;
};

/**
 * Generate Y axis props, ticks, tick formatting.
 *
 * @param {object} params
 * @param {Array} params.dataSets
 * @param {number|object} params.maxY
 * @param {object} params.yAxisPropDefaults
 * @param {Function} params.yAxisTickFormat
 * @returns {Array}
 */
const generateYAxisProps = ({ dataSets = [], maxY, yAxisPropDefaults = {}, yAxisTickFormat } = {}) => {
  const axisProps = [];
  const isMultiAxis = dataSets.length > 1;

  dataSets.forEach(({ id, stroke, strokeWidth } = {}, index) => {
    const updatedAxisProps = {
      style: { axis: {}, tickLabels: {} },
      tickFormat: tick => tick
    };

    if (isMultiAxis && stroke) {
      updatedAxisProps.style.axis.stroke = stroke;
    }

    if (isMultiAxis && strokeWidth) {
      updatedAxisProps.style.axis.strokeWidth = strokeWidth;
    }

    if (typeof yAxisTickFormat === 'function') {
      const updatedMaxY = (typeof maxY === 'number' && maxY) || maxY?.[id];

      updatedAxisProps.tickFormat = tick => {
        const normalizedTick = (isMultiAxis && tick * updatedMaxY) || tick;

        return yAxisTickFormat({
          tick: normalizedTick,
          isMultiAxis,
          maxY: updatedMaxY
        });
      };
    }

    axisProps.push({
      ...yAxisPropDefaults,
      ...updatedAxisProps,
      orientation: (index === 0 && 'left') || 'right'
    });
  });

  return axisProps;
};

/**
 * Generate x,y props.
 *
 * @param {object} params
 * @param {Array} params.dataSets
 * @param {object} params.individualMaxY
 * @param {number} params.maxX
 * @param {number} params.maxY
 * @param {boolean} params.xAxisFixLabelOverlap
 * @param {number} params.xAxisLabelIncrement
 * @param {Function} params.xAxisTickFormat
 * @param {Function} params.yAxisTickFormat
 * @returns {{xAxisProps: object, yAxisProps: Array}}
 */
const generateAxisProps = ({
  dataSets = [],
  individualMaxY = {},
  maxX,
  maxY,
  xAxisFixLabelOverlap = true,
  xAxisLabelIncrement = 1,
  xAxisTickFormat,
  yAxisTickFormat
} = {}) => {
  const xAxisPropDefaults = {
    fixLabelOverlap: xAxisFixLabelOverlap
  };

  const yAxisPropDefaults = {
    dependentAxis: true,
    showGrid: true
  };

  let yAxisDataSets = [];
  let xAxisDataSet;

  dataSets.forEach(dataSet => {
    if (dataSet.yAxisUseDataSet) {
      yAxisDataSets.push(dataSet);
    }
    if (dataSet.xAxisUseDataSet) {
      xAxisDataSet = dataSet;
    }
  });

  if (!yAxisDataSets.length) {
    yAxisDataSets.push(dataSets?.[0]);
  } else {
    yAxisDataSets = yAxisDataSets.slice(0, 2);
  }

  if (!xAxisDataSet) {
    xAxisDataSet = dataSets?.[0] || [];
  }

  const updatedMaxY = (yAxisDataSets.length > 1 && individualMaxY) || maxY;

  return {
    xAxisProps: generateXAxisProps({
      dataSet: xAxisDataSet,
      maxX,
      xAxisLabelIncrement,
      xAxisPropDefaults,
      xAxisTickFormat
    }),
    yAxisProps: generateYAxisProps({ dataSets: yAxisDataSets, maxY: updatedMaxY, yAxisPropDefaults, yAxisTickFormat })
  };
};

const chartHelpers = {
  generateAxisProps,
  generateDomains,
  generateElementsProps,
  generateMaxXY,
  generateTooltipData,
  generateXAxisProps,
  generateYAxisProps
};

export {
  chartHelpers as default,
  chartHelpers,
  generateAxisProps,
  generateDomains,
  generateElementsProps,
  generateMaxXY,
  generateTooltipData,
  generateXAxisProps,
  generateYAxisProps
};
