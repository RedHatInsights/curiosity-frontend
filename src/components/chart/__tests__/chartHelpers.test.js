import { chartHelpers } from '../chartHelpers';

describe('ChartHelpers', () => {
  it('should return specific properties', () => {
    expect(chartHelpers).toMatchSnapshot('specific properties');
  });

  it('should generate a max x, max y from datasets', () => {
    const options = {
      dataSets: [
        {
          id: 'lorem',
          data: [
            { x: 0, y: 10 },
            { x: 1, y: 10 },
            { x: 2, y: 10 },
            { x: 3, y: 10 },
            { x: 4, y: 10 },
            { x: 5, y: 10 }
          ]
        },
        {
          id: 'ipsum',
          isStacked: true,
          data: [
            { x: 0, y: 20 },
            { x: 1, y: 10 },
            { x: 2, y: 30 },
            { x: 3, y: 50 },
            { x: 4, y: 100 },
            { x: 5, y: 10 }
          ]
        },
        {
          id: 'dolor',
          isStacked: true,
          data: [
            { x: 0, y: 20 },
            { x: 1, y: 10 },
            { x: 2, y: 30 },
            { x: 3, y: 50 },
            { x: 4, y: 100 },
            { x: 5, y: 10 }
          ]
        }
      ]
    };

    expect(chartHelpers.generateMaxXY(options)).toMatchSnapshot('max x, y');
  });

  it('should generate domain ranges', () => {
    const options = {
      maxY: 200
    };

    expect(chartHelpers.generateDomains(options)).toMatchSnapshot('domain, maxY number');

    options.maxY = {
      lorem: 10,
      ipsum: 100,
      dolor: 100
    };
    expect(chartHelpers.generateDomains(options)).toMatchSnapshot('domain, maxY object');
  });

  it('should generate element props', () => {
    const options = {
      dataSets: [
        {
          id: 'lorem',
          data: [
            { x: 0, y: 10 },
            { x: 1, y: 10 },
            { x: 2, y: 10 },
            { x: 3, y: 10 },
            { x: 4, y: 10 },
            { x: 5, y: 10 }
          ]
        },
        {
          id: 'ipsum',
          isStacked: true,
          data: [
            { x: 0, y: 20 },
            { x: 1, y: 10 },
            { x: 2, y: 30 },
            { x: 3, y: 50 },
            { x: 4, y: 100 },
            { x: 5, y: 10 }
          ]
        },
        {
          id: 'dolor',
          isStacked: true,
          interpolation: 'lorem ipsum interpolation',
          animate: true,
          data: [
            { x: 0, y: 20 },
            { x: 1, y: 10 },
            { x: 2, y: 30 },
            { x: 3, y: 50 },
            { x: 4, y: 100 },
            { x: 5, y: 10 }
          ]
        }
      ],
      maxX: 6,
      maxY: 200
    };

    expect(chartHelpers.generateElementsProps(options)).toMatchSnapshot('props');
  });

  it('should generate tooltip data/content', () => {
    const mockContent = jest.fn();
    const options = {
      content: mockContent,
      dataSets: [
        {
          id: 'lorem',
          data: [
            { x: 0, y: 10 },
            { x: 1, y: 10 },
            { x: 2, y: 10 },
            { x: 3, y: 10 },
            { x: 4, y: 10 },
            { x: 5, y: 10 }
          ]
        },
        {
          id: 'ipsum',
          isStacked: true,
          data: [
            { x: 0, y: 20 },
            { x: 1, y: 10 },
            { x: 2, y: 30 },
            { x: 3, y: 50 },
            { x: 4, y: 100 },
            { x: 5, y: 10 }
          ]
        },
        {
          id: 'dolor',
          isStacked: true,
          data: [
            { x: 0, y: 20 },
            { x: 1, y: 10 },
            { x: 2, y: 30 },
            { x: 3, y: 50 },
            { x: 4, y: 100 },
            { x: 5, y: 10 }
          ]
        }
      ]
    };

    expect(chartHelpers.generateTooltipData(options)).toMatchSnapshot('tooltip, settings');
    expect(mockContent.mock.calls).toMatchSnapshot('tooltip, content');
  });

  it('should generate x axis props', () => {
    const mockFormat = jest.fn();
    const options = {
      xAxisTickFormat: mockFormat,
      maxX: 6,
      xAxisLabelIncrement: 1,
      dataSet: {
        id: 'lorem',
        data: [
          { x: 0, y: 10 },
          { x: 1, y: 10 },
          { x: 2, y: 10 },
          { x: 3, y: 10 },
          { x: 4, y: 10 },
          { x: 5, y: 10 }
        ]
      }
    };

    const generatedProps = chartHelpers.generateXAxisProps(options);
    expect(generatedProps).toMatchSnapshot('props');

    generatedProps.tickFormat(1);
    expect(mockFormat.mock.calls[0]).toMatchSnapshot('format callback');
  });

  it('should generate y axis props', () => {
    const mockFormat = jest.fn();
    const options = {
      yAxisTickFormat: mockFormat,
      maxY: {
        lorem: 10,
        ipsum: 100
      },
      dataSets: [
        {
          id: 'lorem',
          data: [
            { x: 0, y: 10 },
            { x: 1, y: 10 },
            { x: 2, y: 10 },
            { x: 3, y: 10 },
            { x: 4, y: 10 },
            { x: 5, y: 10 }
          ]
        },
        {
          id: 'ipsum',
          stroke: 'lorem ipsum stroke',
          strokeWidth: 2,
          data: [
            { x: 0, y: 20 },
            { x: 1, y: 10 },
            { x: 2, y: 30 },
            { x: 3, y: 50 },
            { x: 4, y: 100 },
            { x: 5, y: 10 }
          ]
        }
      ]
    };

    const generatedProps = chartHelpers.generateYAxisProps(options);
    expect(generatedProps).toMatchSnapshot('props');

    generatedProps[0].tickFormat(1);
    expect(mockFormat.mock.calls[0]).toMatchSnapshot('format callback, 0, left axis');

    generatedProps[1].tickFormat(1);
    expect(mockFormat.mock.calls[1]).toMatchSnapshot('format callback, 1, right axis');
  });

  it('should aggregated generated x,y axis props', () => {
    const options = {};

    expect(chartHelpers.generateAxisProps(options)).toMatchSnapshot('default settings');
  });
});
