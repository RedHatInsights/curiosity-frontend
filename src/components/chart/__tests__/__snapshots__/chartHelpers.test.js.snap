// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartHelpers should aggregated generated x,y axis props: default settings 1`] = `
Object {
  "xAxisProps": Object {
    "fixLabelOverlap": true,
    "tickFormat": [Function],
    "tickValues": Array [],
  },
  "yAxisProps": Array [
    Object {
      "dependentAxis": true,
      "orientation": "left",
      "showGrid": true,
      "style": Object {
        "axis": Object {},
        "tickLabels": Object {},
      },
      "tickFormat": [Function],
    },
  ],
}
`;

exports[`ChartHelpers should generate a max x, max y from datasets: max x, y 1`] = `
Object {
  "individualMaxY": Object {
    "dolor": 100,
    "ipsum": 100,
    "lorem": 10,
  },
  "maxX": 6,
  "maxY": 200,
}
`;

exports[`ChartHelpers should generate domain ranges: domain, maxY number 1`] = `
Object {
  "domain": Object {
    "y": Array [
      0,
      300,
    ],
  },
}
`;

exports[`ChartHelpers should generate domain ranges: domain, maxY object 1`] = `
Object {
  "domain": Object {
    "y": Array [
      0,
      1.25,
    ],
  },
}
`;

exports[`ChartHelpers should generate element props: props 1`] = `
Object {
  "elements": Array [
    Object {
      "chartType": undefined,
      "props": Object {
        "data": Array [
          Object {
            "x": 0,
            "y": 10,
          },
          Object {
            "x": 1,
            "y": 10,
          },
          Object {
            "x": 2,
            "y": 10,
          },
          Object {
            "x": 3,
            "y": 10,
          },
          Object {
            "x": 4,
            "y": 10,
          },
          Object {
            "x": 5,
            "y": 10,
          },
        ],
        "key": "chart-lorem-",
        "name": "chart-lorem-",
        "style": Object {
          "data": Object {},
        },
        "themeColor": undefined,
        "themeVariant": undefined,
        "x": undefined,
        "y": [Function],
      },
    },
  ],
  "elementsById": Object {
    "lorem": Object {
      "chartType": undefined,
      "props": Object {
        "data": Array [
          Object {
            "x": 0,
            "y": 10,
          },
          Object {
            "x": 1,
            "y": 10,
          },
          Object {
            "x": 2,
            "y": 10,
          },
          Object {
            "x": 3,
            "y": 10,
          },
          Object {
            "x": 4,
            "y": 10,
          },
          Object {
            "x": 5,
            "y": 10,
          },
        ],
        "key": "chart-lorem-",
        "name": "chart-lorem-",
        "style": Object {
          "data": Object {},
        },
        "themeColor": undefined,
        "themeVariant": undefined,
        "x": undefined,
        "y": [Function],
      },
    },
  },
  "stackedElements": Array [
    Object {
      "chartType": undefined,
      "props": Object {
        "data": Array [
          Object {
            "x": 0,
            "y": 20,
          },
          Object {
            "x": 1,
            "y": 10,
          },
          Object {
            "x": 2,
            "y": 30,
          },
          Object {
            "x": 3,
            "y": 50,
          },
          Object {
            "x": 4,
            "y": 100,
          },
          Object {
            "x": 5,
            "y": 10,
          },
        ],
        "key": "chart-ipsum-",
        "name": "chart-ipsum-",
        "style": Object {
          "data": Object {},
        },
        "themeColor": undefined,
        "themeVariant": undefined,
        "x": undefined,
        "y": [Function],
      },
    },
    Object {
      "chartType": undefined,
      "props": Object {
        "animate": true,
        "data": Array [
          Object {
            "x": 0,
            "y": 20,
          },
          Object {
            "x": 1,
            "y": 10,
          },
          Object {
            "x": 2,
            "y": 30,
          },
          Object {
            "x": 3,
            "y": 50,
          },
          Object {
            "x": 4,
            "y": 100,
          },
          Object {
            "x": 5,
            "y": 10,
          },
        ],
        "interpolation": "lorem ipsum interpolation",
        "key": "chart-dolor-",
        "name": "chart-dolor-",
        "style": Object {
          "data": Object {},
        },
        "themeColor": undefined,
        "themeVariant": undefined,
        "x": undefined,
        "y": [Function],
      },
    },
  ],
  "stackedElementsById": Object {
    "dolor": Object {
      "chartType": undefined,
      "props": Object {
        "animate": true,
        "data": Array [
          Object {
            "x": 0,
            "y": 20,
          },
          Object {
            "x": 1,
            "y": 10,
          },
          Object {
            "x": 2,
            "y": 30,
          },
          Object {
            "x": 3,
            "y": 50,
          },
          Object {
            "x": 4,
            "y": 100,
          },
          Object {
            "x": 5,
            "y": 10,
          },
        ],
        "interpolation": "lorem ipsum interpolation",
        "key": "chart-dolor-",
        "name": "chart-dolor-",
        "style": Object {
          "data": Object {},
        },
        "themeColor": undefined,
        "themeVariant": undefined,
        "x": undefined,
        "y": [Function],
      },
    },
    "ipsum": Object {
      "chartType": undefined,
      "props": Object {
        "data": Array [
          Object {
            "x": 0,
            "y": 20,
          },
          Object {
            "x": 1,
            "y": 10,
          },
          Object {
            "x": 2,
            "y": 30,
          },
          Object {
            "x": 3,
            "y": 50,
          },
          Object {
            "x": 4,
            "y": 100,
          },
          Object {
            "x": 5,
            "y": 10,
          },
        ],
        "key": "chart-ipsum-",
        "name": "chart-ipsum-",
        "style": Object {
          "data": Object {},
        },
        "themeColor": undefined,
        "themeVariant": undefined,
        "x": undefined,
        "y": [Function],
      },
    },
  },
}
`;

exports[`ChartHelpers should generate tooltip data/content: tooltip, content 1`] = `
Array [
  Array [
    Object {
      "datum": Object {
        "index": 0,
        "itemsByKey": Object {
          "dolor": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 0,
              "y": 20,
            },
          },
          "ipsum": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 0,
              "y": 20,
            },
          },
          "lorem": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 0,
              "y": 10,
            },
          },
        },
        "x": 0,
        "y": 10,
      },
    },
  ],
  Array [
    Object {
      "datum": Object {
        "index": 1,
        "itemsByKey": Object {
          "dolor": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 1,
              "y": 10,
            },
          },
          "ipsum": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 1,
              "y": 10,
            },
          },
          "lorem": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 1,
              "y": 10,
            },
          },
        },
        "x": 1,
        "y": 10,
      },
    },
  ],
  Array [
    Object {
      "datum": Object {
        "index": 2,
        "itemsByKey": Object {
          "dolor": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 2,
              "y": 30,
            },
          },
          "ipsum": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 2,
              "y": 30,
            },
          },
          "lorem": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 2,
              "y": 10,
            },
          },
        },
        "x": 2,
        "y": 10,
      },
    },
  ],
  Array [
    Object {
      "datum": Object {
        "index": 3,
        "itemsByKey": Object {
          "dolor": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 3,
              "y": 50,
            },
          },
          "ipsum": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 3,
              "y": 50,
            },
          },
          "lorem": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 3,
              "y": 10,
            },
          },
        },
        "x": 3,
        "y": 10,
      },
    },
  ],
  Array [
    Object {
      "datum": Object {
        "index": 4,
        "itemsByKey": Object {
          "dolor": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 4,
              "y": 100,
            },
          },
          "ipsum": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 4,
              "y": 100,
            },
          },
          "lorem": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 4,
              "y": 10,
            },
          },
        },
        "x": 4,
        "y": 10,
      },
    },
  ],
  Array [
    Object {
      "datum": Object {
        "index": 5,
        "itemsByKey": Object {
          "dolor": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 5,
              "y": 10,
            },
          },
          "ipsum": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 5,
              "y": 10,
            },
          },
          "lorem": Object {
            "chartType": undefined,
            "color": "",
            "data": Object {
              "x": 5,
              "y": 10,
            },
          },
        },
        "x": 5,
        "y": 10,
      },
    },
  ],
]
`;

exports[`ChartHelpers should generate tooltip data/content: tooltip, settings 1`] = `
Object {
  "0": Object {
    "itemsByKey": Object {
      "dolor": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 0,
          "y": 20,
        },
      },
      "ipsum": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 0,
          "y": 20,
        },
      },
      "lorem": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 0,
          "y": 10,
        },
      },
    },
    "tooltip": undefined,
    "x": 0,
    "y": null,
  },
  "1": Object {
    "itemsByKey": Object {
      "dolor": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 1,
          "y": 10,
        },
      },
      "ipsum": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 1,
          "y": 10,
        },
      },
      "lorem": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 1,
          "y": 10,
        },
      },
    },
    "tooltip": undefined,
    "x": 1,
    "y": null,
  },
  "2": Object {
    "itemsByKey": Object {
      "dolor": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 2,
          "y": 30,
        },
      },
      "ipsum": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 2,
          "y": 30,
        },
      },
      "lorem": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 2,
          "y": 10,
        },
      },
    },
    "tooltip": undefined,
    "x": 2,
    "y": null,
  },
  "3": Object {
    "itemsByKey": Object {
      "dolor": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 3,
          "y": 50,
        },
      },
      "ipsum": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 3,
          "y": 50,
        },
      },
      "lorem": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 3,
          "y": 10,
        },
      },
    },
    "tooltip": undefined,
    "x": 3,
    "y": null,
  },
  "4": Object {
    "itemsByKey": Object {
      "dolor": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 4,
          "y": 100,
        },
      },
      "ipsum": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 4,
          "y": 100,
        },
      },
      "lorem": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 4,
          "y": 10,
        },
      },
    },
    "tooltip": undefined,
    "x": 4,
    "y": null,
  },
  "5": Object {
    "itemsByKey": Object {
      "dolor": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 5,
          "y": 10,
        },
      },
      "ipsum": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 5,
          "y": 10,
        },
      },
      "lorem": Object {
        "chartType": undefined,
        "color": "",
        "data": Object {
          "x": 5,
          "y": 10,
        },
      },
    },
    "tooltip": undefined,
    "x": 5,
    "y": null,
  },
}
`;

exports[`ChartHelpers should generate x axis props: format callback 1`] = `
Array [
  Object {
    "item": Object {
      "x": 1,
      "y": 10,
    },
    "maxX": 6,
    "nextItem": Object {
      "x": 2,
      "y": 10,
    },
    "previousItem": Object {
      "x": 0,
      "y": 10,
    },
    "tick": 1,
  },
]
`;

exports[`ChartHelpers should generate x axis props: props 1`] = `
Object {
  "tickFormat": [Function],
  "tickValues": Array [
    0,
    1,
    2,
    3,
    4,
    5,
  ],
}
`;

exports[`ChartHelpers should generate y axis props: format callback, 0, left axis 1`] = `
Array [
  Object {
    "isMultiAxis": true,
    "maxY": 10,
    "tick": 10,
  },
]
`;

exports[`ChartHelpers should generate y axis props: format callback, 1, right axis 1`] = `
Array [
  Object {
    "isMultiAxis": true,
    "maxY": 100,
    "tick": 100,
  },
]
`;

exports[`ChartHelpers should generate y axis props: props 1`] = `
Array [
  Object {
    "orientation": "left",
    "style": Object {
      "axis": Object {},
      "tickLabels": Object {},
    },
    "tickFormat": [Function],
  },
  Object {
    "orientation": "right",
    "style": Object {
      "axis": Object {
        "stroke": "lorem ipsum stroke",
        "strokeWidth": 2,
      },
      "tickLabels": Object {},
    },
    "tickFormat": [Function],
  },
]
`;

exports[`ChartHelpers should return specific properties: specific properties 1`] = `
Object {
  "generateAxisProps": [Function],
  "generateDomains": [Function],
  "generateElementsProps": [Function],
  "generateMaxXY": [Function],
  "generateTooltipData": [Function],
  "generateXAxisProps": [Function],
  "generateYAxisProps": [Function],
}
`;
